// @massalabs/massa-web3/dist/cmd/generated/grpc/PublicServiceClientPb.js
"use strict";
/**
 * @fileoverview gRPC-Web generated client stub for massa.api.v1
 * @enhanceable
 * @public
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublicServiceClient = void 0;
const tslib_1 = require("tslib");
// Code generated by protoc-gen-grpc-web. DO NOT EDIT.
// versions:
// 	protoc-gen-grpc-web v1.5.0
// 	protoc              v4.23.3
// source: public.proto
/* eslint-disable */
// @ts-nocheck
const grpcWeb = tslib_1.__importStar(require("grpc-web"));
const public_pb = tslib_1.__importStar(require("./public_pb")); // proto import: "public.proto"
class PublicServiceClient {
    client_;
    hostname_;
    credentials_;
    options_;
    constructor(hostname, credentials, options) {
        if (!options)
            options = {};
        if (!credentials)
            credentials = {};
        options['format'] = 'text';
        this.client_ = new grpcWeb.GrpcWebClientBase(options);
        this.hostname_ = hostname.replace(/\/+$/, '');
        this.credentials_ = credentials;
        this.options_ = options;
    }
    methodDescriptorExecuteReadOnlyCall = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/ExecuteReadOnlyCall', grpcWeb.MethodType.UNARY, public_pb.ExecuteReadOnlyCallRequest, public_pb.ExecuteReadOnlyCallResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.ExecuteReadOnlyCallResponse.deserializeBinary);
    executeReadOnlyCall(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/ExecuteReadOnlyCall', request, metadata || {}, this.methodDescriptorExecuteReadOnlyCall, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/ExecuteReadOnlyCall', request, metadata || {}, this.methodDescriptorExecuteReadOnlyCall);
    }
    methodDescriptorGetBlocks = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetBlocks', grpcWeb.MethodType.UNARY, public_pb.GetBlocksRequest, public_pb.GetBlocksResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetBlocksResponse.deserializeBinary);
    getBlocks(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetBlocks', request, metadata || {}, this.methodDescriptorGetBlocks, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetBlocks', request, metadata || {}, this.methodDescriptorGetBlocks);
    }
    methodDescriptorGetDatastoreEntries = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetDatastoreEntries', grpcWeb.MethodType.UNARY, public_pb.GetDatastoreEntriesRequest, public_pb.GetDatastoreEntriesResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetDatastoreEntriesResponse.deserializeBinary);
    getDatastoreEntries(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetDatastoreEntries', request, metadata || {}, this.methodDescriptorGetDatastoreEntries, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetDatastoreEntries', request, metadata || {}, this.methodDescriptorGetDatastoreEntries);
    }
    methodDescriptorGetEndorsements = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetEndorsements', grpcWeb.MethodType.UNARY, public_pb.GetEndorsementsRequest, public_pb.GetEndorsementsResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetEndorsementsResponse.deserializeBinary);
    getEndorsements(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetEndorsements', request, metadata || {}, this.methodDescriptorGetEndorsements, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetEndorsements', request, metadata || {}, this.methodDescriptorGetEndorsements);
    }
    methodDescriptorGetNextBlockBestParents = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetNextBlockBestParents', grpcWeb.MethodType.UNARY, public_pb.GetNextBlockBestParentsRequest, public_pb.GetNextBlockBestParentsResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetNextBlockBestParentsResponse.deserializeBinary);
    getNextBlockBestParents(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetNextBlockBestParents', request, metadata || {}, this.methodDescriptorGetNextBlockBestParents, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetNextBlockBestParents', request, metadata || {}, this.methodDescriptorGetNextBlockBestParents);
    }
    methodDescriptorGetOperations = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetOperations', grpcWeb.MethodType.UNARY, public_pb.GetOperationsRequest, public_pb.GetOperationsResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetOperationsResponse.deserializeBinary);
    getOperations(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetOperations', request, metadata || {}, this.methodDescriptorGetOperations, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetOperations', request, metadata || {}, this.methodDescriptorGetOperations);
    }
    methodDescriptorGetScExecutionEvents = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetScExecutionEvents', grpcWeb.MethodType.UNARY, public_pb.GetScExecutionEventsRequest, public_pb.GetScExecutionEventsResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetScExecutionEventsResponse.deserializeBinary);
    getScExecutionEvents(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetScExecutionEvents', request, metadata || {}, this.methodDescriptorGetScExecutionEvents, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetScExecutionEvents', request, metadata || {}, this.methodDescriptorGetScExecutionEvents);
    }
    methodDescriptorGetSelectorDraws = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetSelectorDraws', grpcWeb.MethodType.UNARY, public_pb.GetSelectorDrawsRequest, public_pb.GetSelectorDrawsResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetSelectorDrawsResponse.deserializeBinary);
    getSelectorDraws(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetSelectorDraws', request, metadata || {}, this.methodDescriptorGetSelectorDraws, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetSelectorDraws', request, metadata || {}, this.methodDescriptorGetSelectorDraws);
    }
    methodDescriptorGetStakers = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetStakers', grpcWeb.MethodType.UNARY, public_pb.GetStakersRequest, public_pb.GetStakersResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetStakersResponse.deserializeBinary);
    getStakers(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetStakers', request, metadata || {}, this.methodDescriptorGetStakers, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetStakers', request, metadata || {}, this.methodDescriptorGetStakers);
    }
    methodDescriptorGetStatus = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetStatus', grpcWeb.MethodType.UNARY, public_pb.GetStatusRequest, public_pb.GetStatusResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetStatusResponse.deserializeBinary);
    getStatus(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetStatus', request, metadata || {}, this.methodDescriptorGetStatus, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetStatus', request, metadata || {}, this.methodDescriptorGetStatus);
    }
    methodDescriptorGetTransactionsThroughput = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetTransactionsThroughput', grpcWeb.MethodType.UNARY, public_pb.GetTransactionsThroughputRequest, public_pb.GetTransactionsThroughputResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetTransactionsThroughputResponse.deserializeBinary);
    getTransactionsThroughput(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetTransactionsThroughput', request, metadata || {}, this.methodDescriptorGetTransactionsThroughput, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetTransactionsThroughput', request, metadata || {}, this.methodDescriptorGetTransactionsThroughput);
    }
    methodDescriptorQueryState = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/QueryState', grpcWeb.MethodType.UNARY, public_pb.QueryStateRequest, public_pb.QueryStateResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.QueryStateResponse.deserializeBinary);
    queryState(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/QueryState', request, metadata || {}, this.methodDescriptorQueryState, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/QueryState', request, metadata || {}, this.methodDescriptorQueryState);
    }
    methodDescriptorSearchBlocks = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/SearchBlocks', grpcWeb.MethodType.UNARY, public_pb.SearchBlocksRequest, public_pb.SearchBlocksResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.SearchBlocksResponse.deserializeBinary);
    searchBlocks(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/SearchBlocks', request, metadata || {}, this.methodDescriptorSearchBlocks, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/SearchBlocks', request, metadata || {}, this.methodDescriptorSearchBlocks);
    }
    methodDescriptorSearchEndorsements = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/SearchEndorsements', grpcWeb.MethodType.UNARY, public_pb.SearchEndorsementsRequest, public_pb.SearchEndorsementsResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.SearchEndorsementsResponse.deserializeBinary);
    searchEndorsements(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/SearchEndorsements', request, metadata || {}, this.methodDescriptorSearchEndorsements, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/SearchEndorsements', request, metadata || {}, this.methodDescriptorSearchEndorsements);
    }
    methodDescriptorSearchOperations = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/SearchOperations', grpcWeb.MethodType.UNARY, public_pb.SearchOperationsRequest, public_pb.SearchOperationsResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.SearchOperationsResponse.deserializeBinary);
    searchOperations(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/SearchOperations', request, metadata || {}, this.methodDescriptorSearchOperations, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/SearchOperations', request, metadata || {}, this.methodDescriptorSearchOperations);
    }
    methodDescriptorGetOperationABICallStacks = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetOperationABICallStacks', grpcWeb.MethodType.UNARY, public_pb.GetOperationABICallStacksRequest, public_pb.GetOperationABICallStacksResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetOperationABICallStacksResponse.deserializeBinary);
    getOperationABICallStacks(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetOperationABICallStacks', request, metadata || {}, this.methodDescriptorGetOperationABICallStacks, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetOperationABICallStacks', request, metadata || {}, this.methodDescriptorGetOperationABICallStacks);
    }
    methodDescriptorGetSlotABICallStacks = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetSlotABICallStacks', grpcWeb.MethodType.UNARY, public_pb.GetSlotABICallStacksRequest, public_pb.GetSlotABICallStacksResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetSlotABICallStacksResponse.deserializeBinary);
    getSlotABICallStacks(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetSlotABICallStacks', request, metadata || {}, this.methodDescriptorGetSlotABICallStacks, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetSlotABICallStacks', request, metadata || {}, this.methodDescriptorGetSlotABICallStacks);
    }
    methodDescriptorGetSlotTransfers = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/GetSlotTransfers', grpcWeb.MethodType.UNARY, public_pb.GetSlotTransfersRequest, public_pb.GetSlotTransfersResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.GetSlotTransfersResponse.deserializeBinary);
    getSlotTransfers(request, metadata, callback) {
        if (callback !== undefined) {
            return this.client_.rpcCall(this.hostname_ +
                '/massa.api.v1.PublicService/GetSlotTransfers', request, metadata || {}, this.methodDescriptorGetSlotTransfers, callback);
        }
        return this.client_.unaryCall(this.hostname_ +
            '/massa.api.v1.PublicService/GetSlotTransfers', request, metadata || {}, this.methodDescriptorGetSlotTransfers);
    }
    methodDescriptorNewBlocksServer = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/NewBlocksServer', grpcWeb.MethodType.SERVER_STREAMING, public_pb.NewBlocksServerRequest, public_pb.NewBlocksServerResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.NewBlocksServerResponse.deserializeBinary);
    newBlocksServer(request, metadata) {
        return this.client_.serverStreaming(this.hostname_ +
            '/massa.api.v1.PublicService/NewBlocksServer', request, metadata || {}, this.methodDescriptorNewBlocksServer);
    }
    methodDescriptorNewEndorsementsServer = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/NewEndorsementsServer', grpcWeb.MethodType.SERVER_STREAMING, public_pb.NewEndorsementsServerRequest, public_pb.NewEndorsementsServerResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.NewEndorsementsServerResponse.deserializeBinary);
    newEndorsementsServer(request, metadata) {
        return this.client_.serverStreaming(this.hostname_ +
            '/massa.api.v1.PublicService/NewEndorsementsServer', request, metadata || {}, this.methodDescriptorNewEndorsementsServer);
    }
    methodDescriptorNewFilledBlocksServer = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/NewFilledBlocksServer', grpcWeb.MethodType.SERVER_STREAMING, public_pb.NewFilledBlocksServerRequest, public_pb.NewFilledBlocksServerResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.NewFilledBlocksServerResponse.deserializeBinary);
    newFilledBlocksServer(request, metadata) {
        return this.client_.serverStreaming(this.hostname_ +
            '/massa.api.v1.PublicService/NewFilledBlocksServer', request, metadata || {}, this.methodDescriptorNewFilledBlocksServer);
    }
    methodDescriptorNewOperationsServer = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/NewOperationsServer', grpcWeb.MethodType.SERVER_STREAMING, public_pb.NewOperationsServerRequest, public_pb.NewOperationsServerResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.NewOperationsServerResponse.deserializeBinary);
    newOperationsServer(request, metadata) {
        return this.client_.serverStreaming(this.hostname_ +
            '/massa.api.v1.PublicService/NewOperationsServer', request, metadata || {}, this.methodDescriptorNewOperationsServer);
    }
    methodDescriptorNewSlotExecutionOutputsServer = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/NewSlotExecutionOutputsServer', grpcWeb.MethodType.SERVER_STREAMING, public_pb.NewSlotExecutionOutputsServerRequest, public_pb.NewSlotExecutionOutputsServerResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.NewSlotExecutionOutputsServerResponse.deserializeBinary);
    newSlotExecutionOutputsServer(request, metadata) {
        return this.client_.serverStreaming(this.hostname_ +
            '/massa.api.v1.PublicService/NewSlotExecutionOutputsServer', request, metadata || {}, this.methodDescriptorNewSlotExecutionOutputsServer);
    }
    methodDescriptorTransactionsThroughputServer = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/TransactionsThroughputServer', grpcWeb.MethodType.SERVER_STREAMING, public_pb.TransactionsThroughputServerRequest, public_pb.TransactionsThroughputServerResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.TransactionsThroughputServerResponse.deserializeBinary);
    transactionsThroughputServer(request, metadata) {
        return this.client_.serverStreaming(this.hostname_ +
            '/massa.api.v1.PublicService/TransactionsThroughputServer', request, metadata || {}, this.methodDescriptorTransactionsThroughputServer);
    }
    methodDescriptorNewTransfersInfoServer = new grpcWeb.MethodDescriptor('/massa.api.v1.PublicService/NewTransfersInfoServer', grpcWeb.MethodType.SERVER_STREAMING, public_pb.NewTransfersInfoServerRequest, public_pb.NewTransfersInfoServerResponse, (request) => {
        return request.serializeBinary();
    }, public_pb.NewTransfersInfoServerResponse.deserializeBinary);
    newTransfersInfoServer(request, metadata) {
        return this.client_.serverStreaming(this.hostname_ +
            '/massa.api.v1.PublicService/NewTransfersInfoServer', request, metadata || {}, this.methodDescriptorNewTransfersInfoServer);
    }
}
exports.PublicServiceClient = PublicServiceClient;
