// @massalabs/massa-web3/dist/cmd/deweb/keys/deweb_site_keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fileChunkKey = exports.fileChunkCountKey = exports.fileMetadataKey = exports.fileKey = exports.fileLocationKey = exports.globalMetadataKey = exports.DEWEB_VERSION_TAG = exports.GLOBAL_METADATA_TAG = exports.FILE_METADATA_TAG = exports.CHUNK_NB_TAG = exports.CHUNK_TAG = exports.FILE_LOCATION_TAG = exports.FILE_TAG = void 0;
const basicElements_1 = require("../../basicElements");
exports.FILE_TAG = (0, basicElements_1.strToBytes)('\x01FILE');
exports.FILE_LOCATION_TAG = (0, basicElements_1.strToBytes)('\x02LOCATION');
exports.CHUNK_TAG = (0, basicElements_1.strToBytes)('\x03CHUNK');
exports.CHUNK_NB_TAG = (0, basicElements_1.strToBytes)('\x04CHUNK_NB');
exports.FILE_METADATA_TAG = (0, basicElements_1.strToBytes)('\x05FM');
exports.GLOBAL_METADATA_TAG = (0, basicElements_1.strToBytes)('\x06GM');
exports.DEWEB_VERSION_TAG = (0, basicElements_1.strToBytes)('\xFFDEWEB_VERSION');
function globalMetadataKey(metadataKey) {
    return new Uint8Array([...exports.GLOBAL_METADATA_TAG, ...metadataKey]);
}
exports.globalMetadataKey = globalMetadataKey;
function fileLocationKey(hashLocation) {
    return new Uint8Array([...exports.FILE_LOCATION_TAG, ...hashLocation]);
}
exports.fileLocationKey = fileLocationKey;
function fileKey(hashLocation) {
    return new Uint8Array([...exports.FILE_TAG, ...hashLocation]);
}
exports.fileKey = fileKey;
function fileMetadataKey(hashLocation, metadataKey = new Uint8Array()) {
    return new Uint8Array([
        ...fileKey(hashLocation),
        ...exports.FILE_METADATA_TAG,
        ...metadataKey,
    ]);
}
exports.fileMetadataKey = fileMetadataKey;
function fileChunkCountKey(hashLocation) {
    return new Uint8Array([...fileKey(hashLocation), ...exports.CHUNK_NB_TAG]);
}
exports.fileChunkCountKey = fileChunkCountKey;
function fileChunkKey(hashLocation, index) {
    return new Uint8Array([
        ...fileKey(hashLocation),
        ...exports.CHUNK_TAG,
        ...basicElements_1.U32.toBytes(index),
    ]);
}
exports.fileChunkKey = fileChunkKey;
