// @massalabs/massa-web3/dist/cmd/provider/grpcProvider/grpcProvider.js
"use strict";
/* eslint-disable @typescript-eslint/naming-convention */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcProvider = void 0;
const grpcPublicProvider_1 = require("./grpcPublicProvider");
const networks_1 = require("../../utils/networks");
const PublicServiceClientPb_1 = require("../../generated/grpc/PublicServiceClientPb");
/**
 * GrpcProvider implements the Provider interface using gRPC for Massa blockchain interactions
 */
class GrpcProvider extends grpcPublicProvider_1.GrpcPublicProvider {
    client;
    url;
    account;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    _providerName = 'Massa GRPC provider';
    /**
     * Creates a new GrpcProvider instance
     * @param url - The gRPC endpoint URL
     * @param account - The account associated with this provider
     */
    constructor(client, url, account) {
        super(client, url);
        this.client = client;
        this.url = url;
        this.account = account;
    }
    static mainnet(account) {
        return account
            ? GrpcProvider.fromGrpcUrl(networks_1.GrpcApiUrl.Mainnet, account)
            : GrpcProvider.fromGrpcUrl(networks_1.GrpcApiUrl.Mainnet);
    }
    static buildnet(account) {
        return account
            ? GrpcProvider.fromGrpcUrl(networks_1.GrpcApiUrl.Buildnet, account)
            : GrpcProvider.fromGrpcUrl(networks_1.GrpcApiUrl.Buildnet);
    }
    static fromGrpcUrl(url, account) {
        if (account) {
            return new GrpcProvider(new PublicServiceClientPb_1.PublicServiceClient(url), url, account);
        }
        return new grpcPublicProvider_1.GrpcPublicProvider(new PublicServiceClientPb_1.PublicServiceClient(url), url);
    }
    /**
     * Gets the address of the associated account
     */
    get address() {
        return this.account.address.toString();
    }
    /**
     * Gets the name of the associated account
     */
    get accountName() {
        return this.account.address.toString();
    }
    /**
     * Gets the name of the provider
     */
    get providerName() {
        return this._providerName;
    }
    /**
     * Executes a read-only smart contract call
     */
    async readSC(params) {
        const caller = params.caller ?? this.address;
        return super.readSC({ ...params, caller });
    }
    /**
     * Retrieves the balance of the associated account
     */
    async balance(final = true) {
        const balance = await super.balanceOf([this.address], final);
        return balance[0].balance;
    }
    /**
     * @deprecated This method cannot be implemented in the GRPC provider.
     * Use another provider or alternative method for signing data.
     */
    /* eslint-disable-next-line class-methods-use-this */
    sign() {
        throw new Error('Method not implemented.');
    }
    /**
     * @deprecated This method cannot be implemented in the GRPC provider.
     * Use another provider or alternative method for buying rolls.
     */
    /* eslint-disable-next-line class-methods-use-this */
    buyRolls() {
        throw new Error('Method not implemented.');
    }
    /**
     * @deprecated This method cannot be implemented in the GRPC provider.
     * Use another provider or alternative method for selling rolls.
     */
    /* eslint-disable-next-line class-methods-use-this */
    sellRolls() {
        throw new Error('Method not implemented.');
    }
    /**
     * @deprecated This method cannot be implemented in the GRPC provider.
     * Use another provider or alternative method for transferring assets.
     */
    /* eslint-disable-next-line class-methods-use-this */
    transfer() {
        throw new Error('Method not implemented.');
    }
    /**
     * @deprecated This method cannot be implemented in the GRPC provider.
     * Use another provider or alternative method for calling smart contracts.
     */
    /* eslint-disable-next-line class-methods-use-this */
    callSC() {
        throw new Error('Method not implemented.');
    }
    /**
     * @deprecated This method cannot be implemented in the GRPC provider.
     * Use another provider or alternative method for executing smart contracts.
     */
    /* eslint-disable-next-line class-methods-use-this */
    executeSC() {
        throw new Error('Method not implemented.');
    }
    /**
     * @deprecated This method cannot be implemented in the GRPC provider.
     * Use another provider or alternative method for deploying smart contracts.
     */
    /* eslint-disable-next-line class-methods-use-this */
    deploySC() {
        throw new Error('Method not implemented.');
    }
}
exports.GrpcProvider = GrpcProvider;
