// @massalabs/massa-web3/dist/esm/deweb/keys/deweb_site_keys.js
import { strToBytes, U32 } from '../../basicElements';
export const FILE_TAG = strToBytes('\x01FILE');
export const FILE_LOCATION_TAG = strToBytes('\x02LOCATION');
export const CHUNK_TAG = strToBytes('\x03CHUNK');
export const CHUNK_NB_TAG = strToBytes('\x04CHUNK_NB');
export const FILE_METADATA_TAG = strToBytes('\x05FM');
export const GLOBAL_METADATA_TAG = strToBytes('\x06GM');
export const DEWEB_VERSION_TAG = strToBytes('\xFFDEWEB_VERSION');
export function globalMetadataKey(metadataKey) {
    return new Uint8Array([...GLOBAL_METADATA_TAG, ...metadataKey]);
}
export function fileLocationKey(hashLocation) {
    return new Uint8Array([...FILE_LOCATION_TAG, ...hashLocation]);
}
export function fileKey(hashLocation) {
    return new Uint8Array([...FILE_TAG, ...hashLocation]);
}
export function fileMetadataKey(hashLocation, metadataKey = new Uint8Array()) {
    return new Uint8Array([
        ...fileKey(hashLocation),
        ...FILE_METADATA_TAG,
        ...metadataKey,
    ]);
}
export function fileChunkCountKey(hashLocation) {
    return new Uint8Array([...fileKey(hashLocation), ...CHUNK_NB_TAG]);
}
export function fileChunkKey(hashLocation, index) {
    return new Uint8Array([
        ...fileKey(hashLocation),
        ...CHUNK_TAG,
        ...U32.toBytes(index),
    ]);
}
